### PR TITLE
compile: turn off fullgraph=True to support llama4

### PR DESCRIPTION
**Pull request by bdhirsh**
This PR + https://github.com/pytorch/pytorch/pull/153384 is enough to get torchtitan running for me with llama4 and compile
```
CONFIG_FILE="./torchtitan/experiments/llama4/train_configs/debug_model.toml" ./run_train.sh --training.compile
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #1182

### CODE REVIEW

File: torchtitan/models/llama3/parallelize_llama.py [hunk 1]
```diff
@@ -304,7 +304,7 @@ def apply_compile(model: nn.Module):
     repeated structure. Alternatively one can compile the whole model (after applying DP).
     """
     for layer_id, transformer_block in model.layers.named_children():
-        transformer_block = torch.compile(transformer_block, fullgraph=True)
```
Comment (fegin) [inline torchtitan/models/llama3/parallelize_llama.py:307]: Can we add a comment/TODO to remind us to turn it back on when issues are resolved?

**Review by fegin — APPROVED** [review]

File: torchtitan/models/llama3/parallelize_llama.py (hunk 1 above)
Comment (bdhirsh) [reply torchtitan/models/llama3/parallelize_llama.py:307]: oh will do. there are two related things here:

(1) grouped_mm support in compile, which torchtitan uses in llama4. I added basic support in core in this PR: https://github.com/pytorch/pytorch/pull/153384

(2) E2E llama4 + compile in torchtitan. The current reason this completely blows up today is that torchtitan's llama4 + FSDP2 integration requires wrapping the MoE layer, which requires installing backward hooks around the MoE layer. Compile does not support compiling backward hooks (we graph break), and so we need to do one of these options:

(a) allow the graph break (turn off `fullgraph=True`)

(b) tweak torchtitan so that instead of compiling each transformer layer, we compile MoE layers separately, and compile the rest of the transformer block layer separately as well.

I also mentioned this to @tianyu-l but calling it out here: (a) is easier to do, so I'm doing it here, but it does have the risk that if any changes are made to core that increases the number of graph breaks in torchtitan, we won't error as loudly (we may see a perf drop instead). (b) is probably better to do at some point, I'm just doing the simpler thing here.

Are folks working on torchtitan interested in running benchmarks for titan + llama4 (with compile on/off?)

Comment (bdhirsh) [reply torchtitan/models/llama3/parallelize_llama.py:307]: @fegin I actually tweaked the PR so that we still fullgraph=True compile the "regular" transformer blocks, and only fullgraph=False the blocks with MoE layers. I think this should reduce the risk we hit regressions, so this may be a reasonable long term solution (when using FSDP2 in torchtitan), as long as we see reasonable perf numbers.

Comment (fegin) [reply torchtitan/models/llama3/parallelize_llama.py:307]: @bdhirsh Thanks for the thorough explanation. The 8GPU integration test is timeout. Since llama4 is not in the integration test, the integrate test issue should not be caused by this PR. I still relaunch the test but feel free to land it.

**Review by tianyu-l — CHANGES_REQUESTED** [review]
> The current reason this completely blows up today is that torchtitan's llama4 + FSDP2 integration requires wrapping the MoE layer, which requires installing backward hooks around the MoE layer.

hmm let's be more careful here. This is only true when EP is used, specifically dp2ep (e.g. in https://github.com/pytorch/torchtitan/pull/732). Currently in Llama 4, EP is not supported yet, which means we are doing homogeneous FSDP2 wrapping to all the transformer blocks only (not to MoE modules). So I suppose the full graph compilation shouldn't be violated. If we set `full_graph=True`, where would it break?

**Comment by bdhirsh** [comment]
Locally, I was seeing that when we compile each transformer block layer, dynamo was trying (and failing) to graph break, because someone was attempting to install backward hooks inside of one of the transformer blocks. If that's surprising to you I can try to find the code that is installing the backward hook.

**Comment by tianyu-l** [comment]
I think the backward hooks are from the auxiliary-loss-free load balancing (https://github.com/pytorch/torchtitan/pull/1114).

The load balancing algorithm would possess a bias term for each expert, based on the number of tokens an expert has seen so far.
1. The single-device algo needs a backward hook to update the bias term after each iteration.
2. For multi-device, we need another backward hook to all-reduce the bias term across all DP ranks, as different DP ranks see different inputs.

Using forward, forward pre, or backward pre hooks would cause conflict with activation checkpointing.

**Comment by tianyu-l** [comment]
cc @soulitzer on this issue
Basically in order to update the bias terms without intruding model code, I needed to use hooks. To not interfere with AC, I had to use full backward hooks, which breaks full-graph compilation.

I wonder if AC supports optionally bypassing some hook computation, even in the full AC mode?

**Comment by soulitzer** [comment]
@tianyu-l Do you have more details on what the conflict with activation checkpointing is?

Without knowing any more context, one guess is that issue that using forward, forward pre, or backward pre hooks causes conflict with activation checkpointing because it would make the original forward computation and recompute see different bias terms.
And the reason the backward post hook is fine is because it would mean the bias term is updated AFTER the recompute has already been done?
(With this explanation though, forward pre hooks should be fine too?)

**Comment by tianyu-l** [comment]
@soulitzer 
This sounds exciting! Let's definitely try it and see if it can help avoid using backward hooks in this problem.

> This would mean that python side effects no longer run twice

how would you expose control to users so that when users want to recompute some forward hooks they can still do so?

> proper handling in-place aten ops is not ready yet, but should be able to be done in a nicer way than in the previous AC

I guess I'm missing some context on the previous AC treatment of in-place ops.
I do wanna flag that in a recent fix, the `_update_expert_bias` hook uses in-place addition to update the buffer `self.expert_bias`.
https://github.com/pytorch/torchtitan/pull/1226

**Comment by soulitzer** [comment]
@tianyu-l 

> those hooks to update buffers in place (https://github.com/pytorch/torchtitan/pull/1226) only happen in forward but not in backward

I see, overall I'd say this is very much in line with what the new AC wants to achieve, and is what is also tricky to achieve in the old AC. It's great to have this as a concrete use case in MOEs!
(To the perspective of AC forward hooks are not any different from any other logic being executed in the forward pass.)

I've just now added some very basic in-place support only tested in eager so far: "If a tensor is mutated, save its latest version. If someone depends on an earlier version, error". 
https://github.com/soulitzer/ac-experimental/commit/c7f495f4eab12af7b62bc4225fa2349979fa98b8#diff-3665d65394f4f58a56a256ad6dd8621c68118d90fe56a19387e251c19cec2d2eR406
There's still things to be done though, e.g. we may want to smarter things in the case of ops like batch norm.

I'll take a look at the compile side next (about to sign off, but should have something early next week)

**Comment by soulitzer** [comment]
Update:

Was able to run `CONFIG_FILE="./torchtitan/experiments/llama4/train_configs/debug_model.toml" ./run_train.sh --training.compile` without error (didn't check numerics or memory though):

- Moved the sync and bias update logic from the backward hook. But instead of forward hooks, I put the changes at the end of the MOE forward. Dynamo was failing to create guards for the forward hooks possibly due to what it closed over. I don't know how important to put this logic in hooks.
- ~~Ran into a possible SAC/ partitioner bug, I have a hack to get around this but still investigating. (EDIT: a possible fix https://github.com/pytorch/pytorch/pull/156420)~~ Should be fixed on the latest commit of new AC
- Currently 'selective' per-op AC is the only supported setting. Ops that do RNG aren't currently allowed to be recomputed by the new AC. In particular the op in question is sdpa (presumably because it does drop out, although we can probably special case if dropout_p=0). It shouldn't be too hard to support this though, its just not as ideal because we have to stash the RNG state for every single RNG op instance and that is a bit more overhead.
- Patched in https://github.com/pytorch/pytorch/pull/155715 (EDIT: patch is now landed to core, although we may try to find another approach once AC2 is upstreamed) (this patch exposes a utility to disable HOPs from erroring when encountering a side effect - something like this is needed to allow mutating the buffer within the HOP)

**Comment by tianyu-l** [comment]
@soulitzer 
Sounds interesting.

Did you run it on single GPU? I ask because I had imagined that the all-reduce below for syncing purpose would break full-graph compile, too
https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/llama4/infra/parallelize.py#L138

But as https://github.com/pytorch/torchtitan/pull/1304 pointed out, the all-reduce needs to be done outside model across gradient accumulation. So eventually we may not need it in the model code.

**Comment by soulitzer** [comment]
> Did you run it on single GPU? I ask because I had imagined that the all-reduce below for syncing purpose would break full-graph compile, too

Hmm I see all_reduce being traced into this dynamo graph: https://www.internalfb.com/phabricator/paste/view/P1845811599
Maybe it is fine :P 

(I ran on 4 gpus, w/ `NGPU=4`)

**Comment by tianyu-l** [comment]
@soulitzer 
FYI
due to the incompatibility of gradient accumulation and expert bias update, originally reported in https://github.com/pytorch/torchtitan/pull/1304 (expert bias update should not happen every forward/backward, but only once per gradient accumulation steps, i.e. per optimizer step)
I'm now doing the expert bias update in an optimizer step pre-hook: https://github.com/pytorch/torchtitan/pull/1324/files#diff-1140122c42085d443d58ed2bc92db130ae36c5c1248b4de03a891317a15e6171R62
So we no longer have the full backward hooks in the model code.

However, there's another potential issue with per op SAC:
Since there is small matmul happening in [MoE.router.gate](https://github.com/pytorch/torchtitan/blob/7d5f3cc698853d2227cf5433776406d0e0345424/torchtitan/experiments/llama4/model/moe.py#L117), it is interfering with the [SAC policy](https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama3/infra/parallelize.py#L273) of saving every other matmul, presumably causing imbalance save/recomputes. See @H-Huang 's [comment](https://github.com/pytorch/torchtitan/pull/1330#discussion_r2164727536).

Do you think we could have a way to filter out the small matmul from the MoE router (and maybe always do recompute?)?

**Comment by soulitzer** [comment]
> I'm now doing the expert bias update in an optimizer step pre-hook: https://github.com/pytorch/torchtitan/pull/1324/files#diff-1140122c42085d443d58ed2bc92db130ae36c5c1248b4de03a891317a15e6171R62
So we no longer have the full backward hooks in the model code.

I see, thanks for the update!

> Since there is small matmul happening in [MoE.router.gate](https://github.com/pytorch/torchtitan/blob/7d5f3cc698853d2227cf5433776406d0e0345424/torchtitan/experiments/llama4/model/moe.py#L117), it is interfering with the [SAC policy](https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama3/infra/parallelize.py#L273) of saving every other matmul, presumably causing imbalance save/recomputes. See @H-Huang 's https://github.com/pytorch/torchtitan/pull/1330#discussion_r2164727536.

One way of doing this is might be to inspect the shape of the input tensors and ignore the operator in the every-other-matmul policy if the inputs are sufficiently small. How brittle this strategy is I guess depends on whether we care specifically that the skipped matmuls are specifically the gating matmuls and only using size as a proxy.
